bob-apps/bin/bobc.rs

```diff
@@ -1,137 +1,421 @@
+        let file = entry.path();
+            let name = file.to_str().unwrap().to_owned();
+            if let (false, Some(cap)) = (file.is_dir(), re_path.captures(&name)) {
+                let key = cap[1].parse().unwrap();
+                keys_names.push(KeyName { key, name });
+            }
+    }
+    Box::new(keys_names.into_iter())
+}
 
-    let res = client.put(put_req).await;
-    info!("{:#?}", res);
+fn prepare_get(filenames: Box<dyn Iterator<Item = (u64, String)>>) -> Box<dyn Iterator<Item = KeyName>> {
+    Box::new(filenames.map(|(key, name)| KeyName {
+        key,
+        name
+    }))
 }
 
-async fn get(key: Vec<u8>, addr: Uri) {
-    let mut client = BobApiClient::connect(addr).await.unwrap();
+async fn put(key: u64, key_size: usize, filename: &str, client: &mut BobApiClient<Channel>) {
+    match fs::read(filename).await {
+        Ok(data) => {
+            let timestamp = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs();
+            let meta = BlobMeta { timestamp };
+            let blob = Blob {
+                data,
+                meta: Some(meta),
+            };
+            let message = PutRequest {
+                key: Some(BlobKey { key: get_key_value(key, key_size) }),
+                data: Some(blob),
+                options: None,
+            };
+            let put_req = Request::new(message);
+
+            match client.put(put_req).await {
+                Ok(_) => info!("key: {}, file: {}", key, filename),
+                Err(e) => error!("key: {}, file: {}, error: {:?}", key, filename, e)
+            }
+        }
+        Err(e) => {
+            error!("key: {}, file: {}, error: {:?}", key, filename, e)
+        }
+    }
+}
 
+async fn get(key: u64, key_size: usize, filename: &str, client: &mut BobApiClient<Channel>) {
     let message = GetRequest {
-        key: Some(BlobKey { key }),
+        key: Some(BlobKey { key: get_key_value(key, key_size) }),
         options: None,
     };
     let get_req = Request::new(message);
     let res = client.get(get_req).await;
     match res {
         Ok(res) => {
             let res: tonic::Response<_> = res;
-            let data: &Blob = res.get_ref();
-            info!("OK");
-            info!("response meta: {:?})", res.metadata());
-            info!("data len: {}, meta: {:?}", data.data.len(), data.meta);
+            let data = res.get_ref();
+            match fs::write(filename, &data.data).await {
+                Err(e) => error!("key: {}, file: {}, error: {:?}", key, filename, e),
+                _ => info!("key: {}, file: {}", key, filename)
+
+            }
         }
-        Err(res) => {
-            info!("ERR");
-            info!("{:?}", res);
+        Err(e) => {
+            error!("key: {}, file: {}, error: {:?}", key, filename, e)
+        }
+    }
+}
+
+async fn exist(keys: Vec<u64>, key_size: usize, client: &mut BobApiClient<Channel>) {
```

Comment:
Once `KeyPattern` supports iteration over it, you can start passing it here instead of vector

Reply:
Done: [f43491b](https://github.com/qoollo/bob/pull/562/commits/f43491b71021282870b5b1e518cec8c73c8ff689)